workspace_host/interface.js

```diff
@@ -235,11 +252,173 @@ watcher.on('unlinkDir', filename => {
+    let running_workspaces;
+    try {
+        running_workspaces = await docker.listContainers();
+    } catch (err) {
+        /* Nothing to do */
+        return;
+    }
+
+    await async.each(running_workspaces, async (container_info) => {
+        if (container_info.Names.length != 1) return;
+        const name = container_info.Names[0].substring(1); /* Remove the preceding forward slash */
+        if (!name.startsWith('workspace-') || db_workspaces_uuid_set.has(name)) return;
+
+        try {
+            /* Try to kill, but don't care if it's already dead */
+            const container = await docker.getContainer(container_info.Id);
+            await container.kill();
+            await container.remove();
+        } catch (err) {
+            logger.error(err);
+        }
+    });
+}
+
+setInterval(async () => {
+    await pruneStoppedContainers();
+    await pruneRunawayContainers();
+}, config.workspaceHostPruneContainersSec * 1000);
+
+/**
+ * Looks up a docker container by the UUID used to launch it.
+ * Throws an exception if the container was not found or if there
+ * are multiple containers with the same UUID (this shouldn't happen?)
+ * @param {string} launch_uuid UUID to search by
+ * @return Dockerode container object
+ */
+async function _getDockerContainerByLaunchUuid(launch_uuid) {
+    try {
+        const containers = await docker.listContainers({
+            filters: `name=workspace-${launch_uuid}`,
+        });
+        return docker.getContainer(containers[0].Id);
+    } catch (err) {
+        throw new Error(`Could not find unique container by launch UUID: ${launch_uuid}`);
+    }
+}
+
+/**
+ * Looks up a workspace object by the workspace id.
+ * This object contains all columns in the 'workspaces' table as well as:
+ * - local_name (container name)
+ * - s3_name (subdirectory name on s3)
+ * @param {integer} workspace_id Workspace ID to search by.
+ * @return {object} Workspace object, as described above.
+ */
+async function _getWorkspaceAsync(workspace_id) {
+    const result = await sqldb.queryOneRowAsync(sql.get_workspace, {workspace_id});
+    const workspace = result.rows[0];
+    workspace.local_name = `workspace-${workspace.launch_uuid}`;
+    workspace.s3_name = `workspace-${workspace.id}`;
+    return workspace;
+}
+
+/**
+ * Allocates and returns an unused port for a workspace.  This will insert the new port into the workspace table.
+ * @param {object} workspace Workspace object, should at least contain an id.
+ * @return {integer} Port that was allocated to the workspace.
+ */
+let _allocateContainerPortLock = false;
+let _allocateContainerPortLockWaiting = [];
+async function _allocateContainerPort(workspace) {
+    /* Lock and unlock functions.  If the lock is free then this will return immediately.
+       If someone already has the lock then this will wait until someone calls unlock() */
+    async function lock() {
```

Comment:
At some point it would be nice to make a `lib/local-lock.js` with these functions in it, maybe in an object so we can do `let _allocateContainerPortLock = new LocalLock()` and then `_allocateContainerPortLock.lock()`, etc.